### ISSUE TITLE
Semantic Permission Bundles

### ISSUE THREAD

**Issue by tomayac** [w3c/permissions#191]
In https://github.com/w3c/deviceorientation/pull/68#discussion_r265404566, @cdumez wrote the following:

> [A]ll I can say is that I find the terms "accelerometer", "gyroscope" and "magnetometer" confusing when it comes to [the [DeviceOrientation Event](https://w3c.github.io/deviceorientation/)] specification. As in it is not obvious which string relates to which event.

Has there been any previous discussion around *specifying a semantic grouping of lower-level permissions to form higher-level permission "bundles"* (in lack of a better term)? For example, for motion sensors access, the permission an app would request could potentially just be `"motion-sensors"` rather than all of `"accelerometer"`, `"gyroscope"`, and `"magnetometer"`?

This would arguably be more intuitive both for *developers* needing motion sensor data (but maybe unaware of the underlying sensor mechanics) as well as for *end users* who would need to make a decision when prompted.

User agents *could* already now present such lower-level access requests grouped (as outlined in https://github.com/w3c/permissions/issues/92#issuecomment-216825056), however, this requires the user agent to recognize that certain permissions belong together (in the code, they could be requested parallelly or serially, or maybe with `navigator.permissions.requestAll()`), and different user agents might handle this differently, which may be either good or bad, depending on one's viewpoint. 

Looking at the [permission registry](https://w3c.github.io/permissions/#permission-registry), another candidate for permission bundles could be (`"camera"` + `"microphone"` = `"camera-microphone"`). Or—maybe more visionary and with a to-be-bikeshed name—also (`"background-fetch"` + `"background-sync"` + `"persistent-storage"` + `"clipboard"` = `"web-application"`).

I reckon that this is somewhat orthogonal to the static method approach `DeviceMotionEvent.requestPermission()` from @cdumez' https://github.com/w3c/deviceorientation/pull/68 in the concrete case, and more aligned with what @reillyeon commented on in https://github.com/w3c/deviceorientation/issues/57#issuecomment-447129020.

(CC: @anssiko)
(Related issue: https://github.com/w3c/permissions/issues/92)

**Comment by reillyeon** [comment]
> So, just to be sure: does that mean that if a user hypothetically grants permission to "accelerometer", they also implicitly grant access to "gyroscope" and "magnetometer"?

Correct, the Chromium implementation treats these permissions as a single bucket.

> If yes, then "motion-sensors" might be appropriate here.

Yes, I think specifying these as a single permission bundle is fine. In theory the threat model for each of them is slightly different but I don't expect that to be easy to explain to users and so browsers are unlikely to differentiate. It only seems more future-proof to keep them separate (and define the mapping to high-level concepts).

> But, if permission to generic sensors is just blanket "granted" without any permission prompt (and there is no way for the user to disable these APIs via UI), then maybe we don't need these permissions at all? (as they relate to this API, not to Permissions Policy)

Chromium grants the "motion-sensors" permission by default and the UI you posted a screenshot above allows the user to change that default from "allow" to "block". Was is missing in the Chromium implementation (tracked by [issue 947112](https://bugs.chromium.org/p/chromium/issues/detail?id=947112)) is an implementation of the requestPermission() methods which would allow a site to request permission to use the "motion-sensors" permission if the user changed the default to "prompt" (which is not currently a selectable option since the API is missing).

**Comment by tomayac** [comment]
A good candidate for such a semantic permissions bundle would be a videoconferencing permission, consisting of camera, microphone, and notifications. I see so many apps in the VC space request those permissions separately, which causes user confusion, but of course is necessary for the VC app to let users talk and be notified of incoming calls. 

I imagine a prompt UI where the permissions would be presented together under a label like “video conferencing” and could be accepted with one click, but where the prompt UI would still allow users to unselect individual permissions they don’t want to grant.

**Comment by jan-ivar** [comment]
I see arguments here for _prompt_ bundling, not necessarily _permission_ bundling.

> A good candidate for such a semantic permissions bundle would be a videoconferencing permission, consisting of camera, microphone, and notifications. I see so many apps in the VC space request those permissions separately

I think you mean you see separate prompts, because sites don't request _permission_ to camera and microphone, they request _camera and microphone_. The prompt-on-use mental model is the antithesis to up-front bundling, and generally better.

Camera and microphone requests can already be bundled in [getUserMedia](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamconstraints).

I think we need to be careful about separating UX from API, and not have the former drive the latter. Specifically, we should first examine whether user agents can solve prompt bundling without requiring changes in API or model. For instance:
```js
const [stream, permission] = await Promise.all([
  navigator.mediaDevices.getUserMedia({video: true, audio: true}),
  Notification.requestPermission()
]);
```
...seems a detectable pattern to any user agent that wishes to experiment with prompt bundling today, even calling it _"video conferencing"_ if they think this is a good idea. [Backward and forward compatible even](https://jan-ivar.github.io/dummy/gum_and_notification.html)!

But even before that, we should ask whether prompt bundling is a good thing in the first place. We already know most people don't read prompts, and that the rate drops the more text is in them. So any bundling here would result in notifications piggybacking on camera and microphone in practice for many users.

Another, perhaps more intuitive solution to the double-prompt problem is to push back on sites' perceived need to request notification permission upfront in the first place. It's a bad time to ask the user who is merely trying to get into the room (maybe a few minutes late even!)

They only care about notifications later when e.g. someone pings them in chat. That's arguably the most relevant time to ask the user permission. If the web platform is pushing sites to ask early, we should perhaps address that instead?

**Comment by tomayac** [comment]
> Specifically, we should first examine whether user agents can solve prompt bundling without requiring changes in API or model. For instance:
>
>  const [stream, permission] = await Promise.all([
>    navigator.mediaDevices.getUserMedia({video: true, audio: true}),
>    Notification.requestPermission()
>  ]);

That’s an interesting idea; paired with `Promise.*()` (like, for example, `Promise.any()`), it could be made as forgiving or not to the denial of individual promises as the app deems necessary. Not sure how feasible it would be to actually detect such patterns in practice, but I’m acknowledgedly not an expert there. It seems not impossible for sure. 

> They only care about notifications later when e.g. someone pings them in chat.

If the site hasn’t asked for permission, some users may never even see that they are being pinged.

**Comment by jan-ivar** [comment]
> Not sure how feasible it would be to actually detect such patterns in practice

Queue a task and count requests? The important part isn't the promises, but that the APIs are invoked on the same task:
```js
navigator.mediaDevices.getUserMedia({video: true, audio: true});
Notification.requestPermission();
```
